readme.txt
Simplecopter

A simple helicopter simulation built with unity3d

changelog.txt
2014-12-24 13:47 +0000
- initial commit (dev branch)

Assets/scripts/buildings.js
#pragma strict




/******************************************************************************/
function Start () {

   // generate some random buildings
   var x : int;
   var z : int;

   for(x=-200; x<200; x+=20) {
      for(z=-50; z<350; z+=20) {

         var height : int = Random.Range(2,80);

         var prefabBuilding = GameObject.CreatePrimitive(PrimitiveType.Cube);
         prefabBuilding.transform.position = Vector3(
            x + Random.Range(-4,4),
            Random.Range(0, height),
            z + Random.Range(-4,4)
         );

         prefabBuilding.transform.localScale += Vector3(
            Random.Range(3,9),
            height,
            Random.Range(6,15)
         );

         prefabBuilding.transform.localRotation.eulerAngles.x = Random.value > 0.5 ? 0 : 90;
         prefabBuilding.transform.localRotation.eulerAngles.y = Random.value > 0.5 ? 0 : 90;

      }
   }



/**
   // generate wall of cubes
   var y : int;
   for(x=-45; x<45; x+=1) {
      for(y=-45; y<45; y+=1) {

         var c = GameObject.CreatePrimitive(PrimitiveType.Cube);

         c.transform.Translate(Vector3(0,0,20));
         c.transform.Translate(Vector3(x,45+y));

         c.transform.localScale = Vector3(0.5, 0.5, 0.5);
      }
   }

/**/

}

Assets/scripts/mechanics.js
#pragma strict


// flight model
public var flightModelSimple : boolean = true;


/**
helicopter specs
force = mass * acceleration
**/
private var bodyMass     : float = 1000;
private var enginePower  : float = 9.81 * bodyMass;     // just enough to hover
private var engineTorque : float = 180 * Mathf.Deg2Rad; // deg/sec to turn around


// swashplate (cyclic control)
private var swashplateFront : Transform;
private var swashplateBack  : Transform;
private var swashplateLeft  : Transform;
private var swashplateRight : Transform;


// swashplate collective (added factor)
private var swashplateCollective    : float =  0;
private var swashplateCollectiveMin : float = -2;
private var swashplateCollectiveMax : float =  2;


// swashplate cyclic (added factor)
private var swashplateCyclicMax : float = 1;


// helo object
private var helo : Transform;


// rotor object
private var rotor : Transform;


// user control
private var heloControl : flightControl;


// helo flight data
private var heloFlightData : flightData;




/******************************************************************************/
function Start() {

   // user control
   heloControl = transform.GetComponent(flightControl);

   // helo flight data
   heloFlightData = transform.GetComponent(flightData);

   // helo is self
   helo = transform;

   // rotor
   rotor = transform.Find("rotor");

   // swashplate
   swashplateFront = transform.Find("rotor/swashplateFront");
   swashplateBack  = transform.Find("rotor/swashplateBack");
   swashplateLeft  = transform.Find("rotor/swashplateLeft");
   swashplateRight = transform.Find("rotor/swashplateRight");
}




/******************************************************************************/
function FixedUpdate() {


   /**
   switch flight model
   **/
   if(Input.GetKeyDown("f")) {
      flightModelSimple = !flightModelSimple;
   }


   /**
   init swashplate points
   these are factors
   they are affected by collective control and cyclic control
   **/
   var swashplatePowerFront : float = 1;
   var swashplatePowerBack  : float = 1;
   var swashplatePowerRight : float = 1;
   var swashplatePowerLeft  : float = 1;




   /**
   collective control
   this raises/lowers the swashplate evenly
   **/
   // collective up, raise swashplate
   if(heloControl.controlCollective > 0) {

      swashplateCollective = Mathf.Clamp(
         swashplateCollective + Time.deltaTime,
         swashplateCollectiveMin,
         swashplateCollectiveMax
      );
   }

   // collective down, lower swashplate
   else if(heloControl.controlCollective < 0) {

      swashplateCollective = Mathf.Clamp(
         swashplateCollective - Time.deltaTime,
         swashplateCollectiveMin,
         swashplateCollectiveMax
      );
   }

   // idle, lower swashplate
   else if(swashplateCollective > 0) {

      swashplateCollective = Mathf.Clamp(
         swashplateCollective - Time.deltaTime,
         0,
         swashplateCollectiveMax
      );
   }

   // idle, raise swashplate
   else if(swashplateCollective < 0) {

      swashplateCollective = Mathf.Clamp(
         swashplateCollective + Time.deltaTime,
         swashplateCollectiveMin,
         0
      );
   }

   // add collective swashplate power
   swashplatePowerFront += swashplateCollective;
   swashplatePowerBack  += swashplateCollective;
   swashplatePowerRight += swashplateCollective;
   swashplatePowerLeft  += swashplateCollective;




   /**
   yaw control
   this is simplified
   we can pretend to use the rotor's torque
   **/
   helo.rigidbody.AddRelativeTorque(Vector3(
      0.0,
      engineTorque * heloControl.controlYaw * Mathf.Abs(heloControl.controlYaw),
      0.0
   ), ForceMode.Acceleration);




   /**
   cyclic control
   **/
   var controlPitch = heloControl.controlPitch;
   var controlRoll  = heloControl.controlRoll;



   /**
   control model, simple
   **/
   if(flightModelSimple) {

      // pitch
      swashplatePowerFront += swashplateCyclicMax * controlPitch;
      swashplatePowerBack  += swashplateCyclicMax * controlPitch * -1;

      // roll
      swashplatePowerLeft  += swashplateCyclicMax * controlRoll;
      swashplatePowerRight += swashplateCyclicMax * controlRoll * -1;
   }




   /**
   control model, advanced
   inputs are squared for finer control
   this is meant for joystick, not mouse
   experimental
   **/
   else {


      /**
      swashplate, cyclic input
      stick position controls swashplate tilt
      **/


      // square
      controlPitch = controlPitch * Mathf.Abs(controlPitch);
      controlRoll  = controlRoll  * Mathf.Abs(controlRoll);


      // pitch
      swashplatePowerFront += swashplateCyclicMax * controlPitch;
      swashplatePowerBack  += swashplateCyclicMax * controlPitch * -1;

      // roll
      swashplatePowerLeft  += swashplateCyclicMax * controlRoll;
      swashplatePowerRight += swashplateCyclicMax * controlRoll * -1;




      /**
      stabilizer
      rudimentary
      **/




      // default stabilizer values, 0.01 per degree
      var stabilizerPitch : float = heloFlightData.pitch / -100;
      var stabilizerRoll  : float = heloFlightData.roll  / -100;


      // square (matching cyclic input)
      stabilizerPitch *= Mathf.Abs(stabilizerPitch);
      stabilizerRoll  *= Mathf.Abs(stabilizerRoll);


      // apply stabilizer correction, pitch
      swashplatePowerFront += stabilizerPitch;
      swashplatePowerBack  += stabilizerPitch * -1;


      // apply stabilizer correction, roll
      swashplatePowerLeft  += stabilizerRoll;
      swashplatePowerRight += stabilizerRoll * -1;
   }




/*
Retreating blade stall

Helicopters are inherently asymetric flying machines. In forward flight, there
will at any moment in time be one blade moving in the direction of flight
(called the advancing blade) and one moving opposite to the direction of flight
(called the retreating blade).

As the helicopter accellerates, the advancing blade's airspeed in its relative
wind increases; conversely, the airspeed of the retreating blade decreases. As
the helicopter goes faster, various parts of the retreating blade may stall and
no longer generate lift. When this happens, the aircraft nose will rise (as the
effect of the lack of lift takes place 90 degress later in the blade's cycle),
possibly violently.
*/




   /**
   lift power: base
   this is an efficiency factor for the engine power
   **/
   var liftPowerFactor : float = 1;





   /**
   lift power: ground effect (cushion)
   + 0% at 5 metres
   +25% at 0 metres
      TODO: apply to all surfaces, not just ground, given sufficient area
   **/
   var cushionFactor : float = 1 - heloFlightData.altitude / 5;
   cushionFactor = Mathf.Clamp(cushionFactor, 0, 1);
   cushionFactor *= cushionFactor; // square
   liftPowerFactor += cushionFactor * 0.25;




   /**
   lift power: effective translational lift
   + 0% at  0m/s
   +25% at 20m/s and beyond (horizontal speed)
   **/
   var etlPowerFactor : float = heloFlightData.speedHorizontal / 20;
   etlPowerFactor = Mathf.Clamp(etlPowerFactor, 0, 1);
   etlPowerFactor *= etlPowerFactor; // square
   liftPowerFactor += etlPowerFactor * 0.25;




   /**
   apply lift power and swashplate action
   **/
   var liftPowerPerPoint : float = enginePower * liftPowerFactor / 4;


   helo.rigidbody.AddForceAtPosition(
      swashplateFront.up * liftPowerPerPoint * swashplatePowerFront,
      swashplateFront.position,
      ForceMode.Force
   );
   helo.rigidbody.AddForceAtPosition(
      swashplateBack.up * liftPowerPerPoint * swashplatePowerBack,
      swashplateBack.position,
      ForceMode.Force
   );
   helo.rigidbody.AddForceAtPosition(
      swashplateLeft.up * liftPowerPerPoint * swashplatePowerLeft,
      swashplateLeft.position,
      ForceMode.Force
   );
   helo.rigidbody.AddForceAtPosition(
      swashplateRight.up * liftPowerPerPoint * swashplatePowerRight,
      swashplateRight.position,
      ForceMode.Force
   );




/**/
   print(""
      +" F: "+swashplatePowerFront
      +" B: "+swashplatePowerBack
      +" L: "+swashplatePowerLeft
      +" R: "+swashplatePowerRight
   );
/**/
}

Assets/scripts/flightData.js
#pragma strict




// helo object
public var helo : Transform;


// data
public var azimuth : float;
public var pitch   : float;
public var roll    : float;

public var altitude : float;

public var speed           : float;
public var speedHorizontal : float;
public var speedVertical   : float;

public var speedPitch : float;
public var speedRoll  : float;
public var speedYaw   : float;

public var fpvHorizontal : float;
public var fpvVertical   : float;




/******************************************************************************/
function Start() {

   // helo is self
   helo = transform;
}




/******************************************************************************/
function FixedUpdate() {


   // velocities
   var heloVelo      = helo.rigidbody.velocity;
   var heloVeloLocal = transform.InverseTransformDirection(heloVelo);
   var heloTorque    = helo.rigidbody.angularVelocity;


   // azimuth (0-360)
   azimuth = helo.localEulerAngles.y % 360;


   // pitch (+/- 90)
   pitch = helo.localRotation.eulerAngles.x;
   pitch = pitch > 90
      ? 360 - pitch
      :   0 - pitch;


   // roll  (+/- 180)
   roll = helo.localRotation.eulerAngles.z;
   roll = roll > 180
      ? 360 - roll
      :   0 - roll;


   // altitude
   altitude = helo.position.y - 0.48;


   // speed (forward)
   // WRONG, this should not depend on pitch
   speed = 999;
   /*
       var locVel = transform.InverseTransformDirection(rigidbody.velocity);
      locVel.z = MovSpeed;
      rigidbody.velocity = transform.TransformDirection(locVel);
   */
   //heloVeloLocal.z;


   // speed (horizontal)
   speedHorizontal = Vector3(heloVelo.x, 0, heloVelo.z).magnitude;


   // speed (vertical)
   speedVertical = heloVelo.y;


   // torque velocity, pitch
   speedPitch = heloTorque.x * -1 * Mathf.Rad2Deg;


   // torque velocity, roll
   speedRoll = heloTorque.z * -1 * Mathf.Rad2Deg;


   // torque velocity, yaw
   speedYaw = heloTorque.y * -1 * Mathf.Rad2Deg;


   /**
   fpv, flight path vector
   shows real direction of the helicopter
   reminder: radians
   return degrees -180/180
   **/
   fpvHorizontal = Mathf.Atan2(heloVeloLocal.x, heloVeloLocal.z) * Mathf.Rad2Deg;
   fpvVertical   = Mathf.Atan2(heloVeloLocal.y, heloVeloLocal.z) * Mathf.Rad2Deg;
}

Assets/scripts/flightControl.js
#pragma strict




// inputMethod
public var useMouse : boolean = true;


// key bindings
private var key_collective_up   : String = "w";
private var key_collective_down : String = "s";
private var key_yaw_right       : String = "d";
private var key_yaw_left        : String = "a";


// controls
public var controlPitchReal  : float = 0;
public var controlRollReal   : float = 0;
public var controlPitch      : float = 0;
public var controlRoll       : float = 0;
public var controlYaw        : float = 0;
public var controlCollective : float = 0;




/******************************************************************************/
function Update() {

   // keep cursor locked
   Screen.lockCursor = true;
}




/******************************************************************************/
function FixedUpdate() {


   // restart the game on enter key
   if(Input.GetKeyDown("r")) {
      Application.LoadLevel(Application.loadedLevel);
   }


   // switch input method
   if(Input.GetKeyDown("i")) {
      useMouse = !useMouse;
   }



   /**
   cyclic controls, mouse emulating joystick
   forward/back = -1/1
   left/right   = -1/1
   **/
   var inputRoll  : float = 0;
   var inputPitch : float = 0;


   // mouse
   if(useMouse) {
      controlRoll  = Mathf.Clamp(Input.GetAxis("Mouse X"),      -1, 1);
      controlPitch = Mathf.Clamp(Input.GetAxis("Mouse Y") * -1, -1, 1);
/*
      inputRoll  = Input.GetAxis("Mouse X");
      inputPitch = Input.GetAxis("Mouse Y") * -1;

      // real values
      controlRollReal  = Mathf.Clamp(controlRollReal  + inputRoll,  -1, 1);
      controlPitchReal = Mathf.Clamp(controlPitchReal + inputPitch, -1, 1);

      // copy real to exposed
      controlRoll  = controlRollReal;
      controlPitch = controlPitchReal;
*/
   }
   // joystick
   else {
      controlRoll  = Input.GetAxis("Horizontal");
      controlPitch = Input.GetAxis("Vertical");
   }




   /**
   yaw
   **/
   // idle
   controlYaw = 0;

   // yaw, right
   if(Input.GetKey(key_yaw_right)) {
      controlYaw = 1;
   }
   // yaw, left
   else if(Input.GetKey(key_yaw_left)) {
      controlYaw = -1;
   }




   /**
   collective control, modifies engine power
   **/
   // idle
   controlCollective = 0;

   // collective up
   if(Input.GetKey(key_collective_up)) {
      controlCollective = 1;
   }
   // collective down
   else if(Input.GetKey(key_collective_down)) {
      controlCollective = -1;
   }
}

Assets/scripts/hud.js
#pragma strict


/**
TODO
   - visual speed indicator (bars) (forward/up/yaw)
   - engine power indicator
**/


// flight data
private var heloFlightData : flightData;


// flight mechanics
private var heloMechanics : mechanics;


// flight mechanics
private var heloFlightControl : flightControl;


// hud elements
private var hud_cam        : Transform;
private var hud_light      : Transform;
private var hud_aoa        : Transform;
private var hud_aoa_level  : Transform;
private var hud_roll_level : Transform;
private var hud_fpv        : Transform;
private var hud_boresight  : Transform;


private var hud_pitch_reader  : TextMesh;
private var hud_azimuth       : TextMesh;
private var hud_roll_reader   : TextMesh;
private var hud_alt_reader    : TextMesh;
private var hud_speed_reader  : TextMesh;
private var hud_vspeed_reader : TextMesh;


private var flightModelText : UI.Text;
private var inputMethodText : UI.Text;


// original position/rotation of hud objects
private var hud_aoa_baseRot        : Quaternion;
private var hud_aoa_level_basePos  : Vector3;
private var hud_roll_level_baseRot : Quaternion;
private var hud_fpv_basePos        : Vector3;


// fpv max angle (stick to screen edges)
private var hud_fpv_angle_max_horizontal : float = 0;
private var hud_fpv_angle_max_vertical   : float = 0;




/******************************************************************************/
function Start() {


   // get mechanics
   heloMechanics = GameObject.Find("helo").transform.GetComponent(mechanics);


   // get flight data
   heloFlightData = GameObject.Find("helo").transform.GetComponent(flightData);


   // get flight control
   heloFlightControl = GameObject.Find("helo").transform.GetComponent(flightControl);


   /**
   set hud camera size depending on main camera fov
   we project relative to the world (pitch levels inaccurate though)
   main square is 90 deg (+/- 45) and 10 units
   1.25 unit covers 45 deg
   **/
   var heloCam = GameObject.Find("helo/cam").GetComponent(Camera);
   var hudCam  = transform.Find("cam").GetComponent(Camera);
   hudCam.orthographicSize = 1.25 * Mathf.Tan(heloCam.fieldOfView / 2 * Mathf.Deg2Rad);


   // get hud objects
   hud_cam        = transform.Find("cam");
   hud_light      = transform.Find("light");
   hud_aoa        = transform.Find("aoa");
   hud_aoa_level  = transform.Find("aoa/level");
   hud_fpv        = transform.Find("fpv");
   hud_boresight  = transform.Find("boresight");
   hud_roll_level = transform.Find("roll/level");


   hud_pitch_reader  = transform.Find("pitch-reader").GetComponent(TextMesh);
   hud_azimuth       = transform.Find("azimuth").GetComponent(TextMesh);
   hud_roll_reader   = transform.Find("roll-reader").GetComponent(TextMesh);
   hud_alt_reader    = transform.Find("alt-reader").GetComponent(TextMesh);
   hud_speed_reader  = transform.Find("speed-reader").GetComponent(TextMesh);
   hud_vspeed_reader = transform.Find("vspeed-reader").GetComponent(TextMesh);


   // get flight model text component
   flightModelText = GameObject.Find("overlayText").Find("flightModel").GetComponent(UI.Text);
   inputMethodText = GameObject.Find("overlayText").Find("inputMethod").GetComponent(UI.Text);


   // original pos/rotation of hud objects
   hud_aoa_baseRot        = hud_aoa.localRotation;
   hud_aoa_level_basePos  = hud_aoa_level.localPosition;
   hud_roll_level_baseRot = hud_roll_level.localRotation;
   hud_fpv_basePos        = hud_fpv.localPosition;


   // fpv angle limit (as tan)
   hud_fpv_angle_max_vertical   = Mathf.Tan(heloCam.fieldOfView/2*Mathf.Deg2Rad);
   hud_fpv_angle_max_horizontal = hud_fpv_angle_max_vertical * heloCam.aspect;
}




/******************************************************************************/
function Update() {


   // update hud text
   hud_azimuth.text       = "AZT\n" + (Mathf.Round(heloFlightData.azimuth) % 360).ToString("000");
   hud_pitch_reader.text  = heloFlightData.pitch.ToString("000.00") + " PITCH";
   hud_roll_reader.text   = heloFlightData.roll.ToString("000.00")  + " ROLL ";
   hud_alt_reader.text    = "ALT "  + heloFlightData.altitude.ToString("0000.00");
   hud_speed_reader.text  = "SPD "  + heloFlightData.speedHorizontal.ToString("000.00");
   hud_vspeed_reader.text = "VSPD " + heloFlightData.speedVertical.ToString("000.00");




   // update flight model text
   flightModelText.text = "flight model:\n"+(heloMechanics.flightModelSimple ? 'simple' : 'advanced');
   inputMethodText.text = "input method:\n"+(heloFlightControl.useMouse ? 'mouse' : 'joystick');



   // update hud aoa level, apply pitch
   // convert rotation to vertical translation, 1unit = 36deg
   hud_aoa_level.localPosition.y =
      hud_aoa_level_basePos.y - (heloFlightData.pitch / 36);




   // update hud aoa, apply roll
   hud_aoa.localRotation.eulerAngles.z =
      hud_aoa_baseRot.eulerAngles.z + heloFlightData.roll;




   // update hud roll, apply roll
   hud_roll_level.localRotation.eulerAngles.y =
      hud_roll_level_baseRot.eulerAngles.y - heloFlightData.roll;




   // fpv
   var fpvH : float = heloFlightData.fpvHorizontal;
   var fpvV : float = heloFlightData.fpvVertical;

   // limit to 90ish degrees
   fpvH = Mathf.Clamp(fpvH, -89.999, 89.999);
   fpvV = Mathf.Clamp(fpvV, -89.999, 89.999);

   // adapt for screen
   fpvH = Mathf.Tan(fpvH * Mathf.Deg2Rad);
   fpvV = Mathf.Tan(fpvV * Mathf.Deg2Rad);

   // screen angle limit (tan)
   fpvH = Mathf.Clamp(
      fpvH,
      hud_fpv_angle_max_horizontal * -1,
      hud_fpv_angle_max_horizontal
   );
   fpvV = Mathf.Clamp(
      fpvV,
      hud_fpv_angle_max_vertical * -1,
      hud_fpv_angle_max_vertical
   );

   // scale for screen
   fpvH *= 1.25;
   fpvV *= 1.25;

   // apply position
   hud_fpv.localPosition = hud_fpv_basePos + Vector3(fpvH, fpvV, 0);




/**
   // debug fpv visually
   var fpv2       = GameObject.Find("helo").Find("fpv2").transform;
   var fpv2visual = fpv2.Find("visual").transform;
   var vl = heloFlightData.helo.InverseTransformDirection(heloFlightData.helo.rigidbody.velocity).normalized;
   fpv2visual.localPosition = vl;
/**
   print(heloFlightData.fpvVertical+"   "+heloFlightData.fpvHorizontal);
/**/
}